modules/dnn/src/layers/harmax.cpp

```diff
@@ -0,0 +1,152 @@
+#include <inttypes.h>
+#include <opencv2/dnn/shape_utils.hpp>
+#include "../precomp.hpp"
+#include "layers_common.hpp"
+#include "../ie_ngraph.hpp"
+
+namespace cv
+{
+namespace dnn
+{
+
+class LayerHardmaxImpl CV_FINAL : public HardmaxLayer
+{
+public:
+    int axis;
+    LayerHardmaxImpl(const LayerParams& params)
+    {
+        axis = params.get<int>("axis", -1);
+    }
+
+
+    virtual bool supportBackend(int backendId) CV_OVERRIDE
+    {
+        return backendId == DNN_BACKEND_OPENCV ||
+               backendId == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH;
+    }
+
+    void getTypes(const std::vector<MatType>& inputs,
+        const int requiredOutputs,
+        const int requiredInternals,
+        std::vector<MatType>& outputs,
+        std::vector<MatType>& internals) const CV_OVERRIDE
+    {
+        CV_Assert(inputs.size());
+        for (auto input : inputs)
+        {
+            if (preferableTarget == DNN_TARGET_OPENCL_FP16)
+                CV_CheckType(input, input == CV_16F || input == CV_8S || input == CV_8U || input == CV_32S || input == CV_64S || input == CV_Bool, "");
+            else
+                CV_CheckType(input, input == CV_32F || input == CV_8S || input == CV_8U || input == CV_32S || input == CV_64S || input == CV_Bool, "");
+        }
+
+        outputs.assign(requiredOutputs, inputs[0]);
+    }
+
+    bool getMemoryShapes(const std::vector<MatShape> &inputs,
+                         const int requiredOutputs,
+                         std::vector<MatShape> &outputs,
+                         std::vector<MatShape> &internals) const CV_OVERRIDE
+    {
+        CV_CheckEQ(inputs.size(), 1ull, "Hardmax: one input is expected");
+        outputs.resize(1);
+        outputs[0] = inputs[0];
+        return true;
```

Comment:
Returning `true` here stands for input and output can be a shared memory, which is not valid for calling `memset` on dst (because it also sets the input zero). So return `false` here.